[flaky] pkg/kncloudevents.TestDispatchMessageToTLSEndpoint

* Test name: kncloudevents.TestDispatchMessageToTLSEndpoint
* Repository: eventing
* Flaky run: https://prow.knative.dev/view/gs/knative-prow/logs/release_eventing_main_periodic/1716789998008668160

```
=== RUN   TestDispatchMessageToTLSEndpoint
    logger.go:130: 2023-10-24T12:29:12.163Z	DEBUG	eventingtls/eventingtls.go:106	Loading key pair	{"tls.secret": "knative-tests/tls-secret"}
    logger.go:130: 2023-10-24T12:29:12.170Z	DEBUG	eventingtls/eventingtls.go:114	certificate stored	{"tls.secret": "knative-tests/tls-secret"}
    event_dispatcher_test.go:962: 
        	Error Trace:	/home/prow/go/src/knative.dev/eventing/pkg/kncloudevents/event_dispatcher_test.go:962
        	Error:      	Expected nil, but got: &fmt.wrapError{msg:"unable to complete request to https://localhost:8334/: Post \"https://localhost:8334/\": dial tcp 127.0.0.1:8334: connect: connection refused", err:(*url.Error)(0xc0007e2630)}
        	Test:       	TestDispatchMessageToTLSEndpoint
--- FAIL: TestDispatchMessageToTLSEndpoint (0.52s)
FAIL pkg/kncloudevents.TestDispatchMessageToTLSEndpoint (0.52s)
```

--- a/pkg/adapter/mtping/runner_test.go
+++ b/pkg/adapter/mtping/runner_test.go
@@ -20,6 +20,7 @@ import (
 	"bytes"
 	"context"
 	"encoding/base64"
+	"fmt"
 	"net/http"
 	"net/http/httptest"
 	"reflect"
@@ -213,8 +214,8 @@ func TestSendEventsTLS(t *testing.T) {
 	eventsChan := make(chan cloudevents.Event, 10)
 	handler := eventingtlstesting.EventChannelHandler(eventsChan)
 	events := make([]cloudevents.Event, 0, 8)
-	ca := eventingtlstesting.StartServer(ctx, t, 8500, handler)
-	hostString := "localhost:8500"
+	ca, port := eventingtlstesting.StartServer(ctx, t, 0, handler)
+	hostString := fmt.Sprintf("localhost:%d", port)
 
 	var wg sync.WaitGroup
 	wg.Add(1)

--- a/pkg/adapter/v2/cloudevents_test.go
+++ b/pkg/adapter/v2/cloudevents_test.go
@@ -18,6 +18,7 @@ package adapter
 
 import (
 	"context"
+	"fmt"
 	nethttp "net/http"
 	"os"
 	"strconv"
@@ -309,7 +310,7 @@ func TestTLS(t *testing.T) {
 	ctx, cancel := context.WithCancel(ctx)
 	t.Cleanup(cancel)
 
-	ca := eventingtlstesting.StartServer(ctx, t, 8333, nethttp.HandlerFunc(func(writer nethttp.ResponseWriter, request *nethttp.Request) {
+	ca, port := eventingtlstesting.StartServer(ctx, t, 0, nethttp.HandlerFunc(func(writer nethttp.ResponseWriter, request *nethttp.Request) {
 		if request.TLS == nil {
 			// It's not on TLS, fail request
 			writer.WriteHeader(nethttp.StatusInternalServerError)
@@ -328,17 +329,17 @@ func TestTLS(t *testing.T) {
 	}{
 		{
 			name:    "https sink URL, no CA certs fail",
-			sink:    "https://localhost:8333",
+			sink:    fmt.Sprintf("https://localhost:%d", port),
 			wantErr: true,
 		},
 		{
 			name:    "https sink URL with ca certs",
-			sink:    "https://localhost:8333",
+			sink:    fmt.Sprintf("https://localhost:%d", port),
 			caCerts: pointer.String(ca),
 		},
 		{
 			name:    "http sink URL with ca certs",
-			sink:    "http://localhost:8333",
+			sink:    fmt.Sprintf("http://localhost:%d", port),
 			caCerts: pointer.String(ca),
 			wantErr: true,
 		},

--- a/pkg/eventingtls/eventingtlstesting/eventingtlstesting.go
+++ b/pkg/eventingtls/eventingtlstesting/eventingtlstesting.go
@@ -47,7 +47,7 @@ func init() {
 	CA, Key, Crt = loadCerts()
 }
 
-func StartServer(ctx context.Context, t *testing.T, port int, handler http.Handler, receiverOptions ...kncloudevents.HTTPEventReceiverOption) string {
+func StartServer(ctx context.Context, t *testing.T, port int, handler http.Handler, receiverOptions ...kncloudevents.HTTPEventReceiverOption) (string, int) {
 	secret := types.NamespacedName{
 		Namespace: "knative-tests",
 		Name:      "tls-secret",
@@ -83,7 +83,7 @@ func StartServer(ctx context.Context, t *testing.T, port int, handler http.Handl
 		}
 	}()
 
-	return string(CA)
+	return string(CA), receiver.GetPort()
 }
 
 func loadCerts() ([]byte, []byte, []byte) {

--- a/pkg/kncloudevents/event_dispatcher_test.go
+++ b/pkg/kncloudevents/event_dispatcher_test.go
@@ -973,9 +973,9 @@ func testEventFormat(t *testing.T, format v1.FormatType, expectedEncoding bindin
 		w.WriteHeader(http.StatusOK)
 	})
 
-	destinationCA := eventingtlstesting.StartServer(ctx, t, port, destinationHandler, kncloudevents.WithDrainQuietPeriod(time.Millisecond))
+	destinationCA, finalPort := eventingtlstesting.StartServer(ctx, t, port, destinationHandler, kncloudevents.WithDrainQuietPeriod(time.Millisecond))
 	destination := duckv1.Addressable{
-		URL:     apis.HTTPS(fmt.Sprintf("localhost:%d", port)),
+		URL:     apis.HTTPS(fmt.Sprintf("localhost:%d", finalPort)),
 		CACerts: &destinationCA,
 	}
 
@@ -1030,9 +1030,9 @@ func TestDispatchMessageToTLSEndpoint(t *testing.T) {
 	destinationEventsChan := make(chan cloudevents.Event, 10)
 	destinationReceivedEvents := make([]cloudevents.Event, 0, 10)
 	destinationHandler := eventingtlstesting.EventChannelHandler(destinationEventsChan)
-	destinationCA := eventingtlstesting.StartServer(ctx, t, 8334, destinationHandler, kncloudevents.WithDrainQuietPeriod(time.Millisecond))
+	destinationCA, port := eventingtlstesting.StartServer(ctx, t, 0, destinationHandler, kncloudevents.WithDrainQuietPeriod(time.Millisecond))
 	destination := duckv1.Addressable{
-		URL:     apis.HTTPS("localhost:8334"),
+		URL:     apis.HTTPS(fmt.Sprintf("localhost:%d", port)),
 		CACerts: &destinationCA,
 	}
 
@@ -1085,9 +1085,9 @@ func TestDispatchMessageToTLSEndpointWithReply(t *testing.T) {
 		w.Write(eventToReply.Data())
 	})
 
-	destinationCA := eventingtlstesting.StartServer(ctxDestination, t, 8335, destinationHandler, kncloudevents.WithDrainQuietPeriod(time.Millisecond))
+	destinationCA, destinationPort := eventingtlstesting.StartServer(ctxDestination, t, 0, destinationHandler, kncloudevents.WithDrainQuietPeriod(time.Millisecond))
 	destination := duckv1.Addressable{
-		URL:     apis.HTTPS("localhost:8335"),
+		URL:     apis.HTTPS(fmt.Sprintf("localhost:%d", destinationPort)),
 		CACerts: &destinationCA,
 	}
 
@@ -1095,9 +1095,9 @@ func TestDispatchMessageToTLSEndpointWithReply(t *testing.T) {
 	replyEventChan := make(chan cloudevents.Event, 10)
 	replyHandler := eventingtlstesting.EventChannelHandler(replyEventChan)
 	replyReceivedEvents := make([]cloudevents.Event, 0, 10)
-	replyCA := eventingtlstesting.StartServer(ctxReply, t, 8336, replyHandler, kncloudevents.WithDrainQuietPeriod(time.Millisecond))
+	replyCA, replyPort := eventingtlstesting.StartServer(ctxReply, t, 0, replyHandler, kncloudevents.WithDrainQuietPeriod(time.Millisecond))
 	reply := duckv1.Addressable{
-		URL:     apis.HTTPS("localhost:8336"),
+		URL:     apis.HTTPS(fmt.Sprintf("localhost:%d", replyPort)),
 		CACerts: &replyCA,
 	}
 
@@ -1145,9 +1145,9 @@ func TestDispatchMessageToTLSEndpointWithDeadLetterSink(t *testing.T) {
 		w.WriteHeader(http.StatusInternalServerError)
 	})
 
-	destinationCA := eventingtlstesting.StartServer(ctxDestination, t, 8337, destinationHandler, kncloudevents.WithDrainQuietPeriod(time.Millisecond))
+	destinationCA, destinationPort := eventingtlstesting.StartServer(ctxDestination, t, 0, destinationHandler, kncloudevents.WithDrainQuietPeriod(time.Millisecond))
 	destination := duckv1.Addressable{
-		URL:     apis.HTTPS("localhost:8337"),
+		URL:     apis.HTTPS(fmt.Sprintf("localhost:%d", destinationPort)),
 		CACerts: &destinationCA,
 	}
 
@@ -1155,9 +1155,9 @@ func TestDispatchMessageToTLSEndpointWithDeadLetterSink(t *testing.T) {
 	dlsEventChan := make(chan cloudevents.Event, 10)
 	dlsHandler := eventingtlstesting.EventChannelHandler(dlsEventChan)
 	dlsReceivedEvents := make([]cloudevents.Event, 0, 10)
-	dlsCA := eventingtlstesting.StartServer(ctxDls, t, 8338, dlsHandler, kncloudevents.WithDrainQuietPeriod(time.Millisecond))
+	dlsCA, dlsPort := eventingtlstesting.StartServer(ctxDls, t, 0, dlsHandler, kncloudevents.WithDrainQuietPeriod(time.Millisecond))
 	dls := duckv1.Addressable{
-		URL:     apis.HTTPS("localhost:8338"),
+		URL:     apis.HTTPS(fmt.Sprintf("localhost:%d", dlsPort)),
 		CACerts: &dlsCA,
 	}
 

--- a/pkg/kncloudevents/event_receiver.go
+++ b/pkg/kncloudevents/event_receiver.go
@@ -22,6 +22,7 @@ import (
 	"fmt"
 	"net"
 	"net/http"
+	"strconv"
 	"time"
 
 	"go.opencensus.io/plugin/ochttp"
@@ -34,7 +35,8 @@ const (
 )
 
 type HTTPEventReceiver struct {
-	port int
+	desiredPort int
+	port        int
 
 	server   *http.Server
 	listener net.Listener
@@ -49,12 +51,13 @@ type HTTPEventReceiver struct {
 // HTTPEventReceiverOption enables further configuration of a HTTPEventReceiver.
 type HTTPEventReceiverOption func(*HTTPEventReceiver)
 
-func NewHTTPEventReceiver(port int, o ...HTTPEventReceiverOption) *HTTPEventReceiver {
+// NewHTTPEventReceiver creates a new HTTPEventReceiver. When desiredPort is set to 0, a free port will be chosen.
+// The final port of the running server will be stored in receiver.Port.
+func NewHTTPEventReceiver(desiredPort int, o ...HTTPEventReceiverOption) *HTTPEventReceiver {
 	h := &HTTPEventReceiver{
-		port: port,
+		desiredPort: desiredPort,
+		Ready:       make(chan interface{}),
 	}
-
-	h.Ready = make(chan interface{})
 
 	for _, opt := range o {
 		opt(h)
@@ -122,10 +125,19 @@ func (recv *HTTPEventReceiver) GetAddr() string {
 	return ""
 }
 
+// GetPort returns the final assigned port of the server.
+// This is blocking, as we need to wait until the server is running.
+func (recv *HTTPEventReceiver) GetPort() int {
+	// wait until server is ready, as only then the port is assigned
+	<-recv.Ready
+
+	return recv.port
+}
+
 // Blocking
 func (recv *HTTPEventReceiver) StartListen(ctx context.Context, handler http.Handler) error {
 	var err error
-	if recv.listener, err = net.Listen("tcp", fmt.Sprintf(":%d", recv.port)); err != nil {
+	if recv.listener, err = net.Listen("tcp", fmt.Sprintf(":%d", recv.desiredPort)); err != nil {
 		return err
 	}
 
@@ -139,6 +151,18 @@ func (recv *HTTPEventReceiver) StartListen(ctx context.Context, handler http.Han
 	}
 	recv.server.Addr = recv.listener.Addr().String()
 	recv.server.Handler = drainer
+
+	_, portStr, err := net.SplitHostPort(recv.server.Addr)
+	if err != nil {
+		return fmt.Errorf("could not get port of server: %w", err)
+	}
+
+	port, err := strconv.Atoi(portStr)
+	if err != nil {
+		return fmt.Errorf("could not convert port to int: %w", err)
+	}
+
+	recv.port = port
 
 	errChan := make(chan error, 1)
 	go func() {

